hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeUpgradeActions.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    public void externalizeFreezeIfUpgradePending() {
+        // TODO: check whether networkCtx.hasPreparedUpgrade() once networkContext is available
+        final var isUpgradePrepared = true; // TODO: isUpgradePrepared = networkCtx.hasPreparedUpgrade();
+
+        if (isUpgradePrepared) {
+            writeCheckMarker(NOW_FROZEN_MARKER);
+        }
+    }
+
+    public CompletableFuture<Void> extractTelemetryUpgrade(final byte[] archiveData, final Instant now) {
+        return extractNow(archiveData, TELEMETRY_UPGRADE_DESC, EXEC_TELEMETRY_MARKER, now);
+    }
+
+    public CompletableFuture<Void> extractSoftwareUpgrade(final byte[] archiveData) {
+        return extractNow(archiveData, PREPARE_UPGRADE_DESC, EXEC_IMMEDIATE_MARKER, null);
+    }
+
+    public void scheduleFreezeOnlyAt(final Instant freezeTime) {
+        withNonNullDualState("schedule freeze", ds -> ds.setFreezeTime(freezeTime));
+    }
+
+    public void scheduleFreezeUpgradeAt(final Instant freezeTime) {
+        withNonNullDualState("schedule freeze", ds -> {
+            ds.setFreezeTime(freezeTime);
+            writeSecondMarker(FREEZE_SCHEDULED_MARKER, freezeTime);
+        });
+    }
+
+    public void abortScheduledFreeze() {
+        withNonNullDualState("abort freeze", ds -> {
+            ds.setFreezeTime(null);
+            writeCheckMarker(FREEZE_ABORTED_MARKER);
+        });
+    }
+
+    public boolean isFreezeScheduled() {
+        final var ans = new AtomicBoolean();
+        withNonNullDualState("check freeze schedule", ds -> {
+            final var freezeTime = ds.getFreezeTime();
+            ans.set(freezeTime != null && !freezeTime.equals(ds.getLastFrozenTime()));
+        });
+        return ans.get();
+    }
+
+    public void catchUpOnMissedSideEffects() {
+        catchUpOnMissedFreezeScheduling();
+        // TODO: fix and then reenable this once networkContext is available
+        //  catchUpOnMissedUpgradePrep();
+    }
+
+    /* --- Internal methods --- */
+
+    private CompletableFuture<Void> extractNow(
+            final byte[] archiveData, final String desc, final String marker, @Nullable final Instant now) {
+        final var size = archiveData.length;
+        final var artifactsLoc = adminServiceConfig.upgradeArtifactsPath();
+        log.info("About to unzip {} bytes for {} update into {}", size, desc, artifactsLoc);
+        return runAsync(() -> {
+            try {
+                FileUtils.cleanDirectory(new File(artifactsLoc));
+                boolean result = UnzipUtility.unzip(archiveData, artifactsLoc);
+                if (!result) throw new IOException("Unzip failed");
+                log.info("Finished unzipping {} bytes for {} update into {}", size, desc, artifactsLoc);
+                writeSecondMarker(marker, now);
+            } catch (final IOException e) {
+                log.error("Failed to unzip archive for NMT consumption", e);
+                log.error(MANUAL_REMEDIATION_ALERT);
```

Comment:
It is correct. The upgrade process checks for the presence or absence of the second marker to know whether the unzip worked correctly. We spawn a separate thread for unzipping because it takes awhile, impacts E2E latency, and we don't actually need it to complete until just before the freeze time. In practice there is always a significant delay on the scale of hours between the time this code is triggered and the freeze time.
I updated the comments in this method to explain this.